### PR TITLE
fix: special asset pairs bugfix and tests

### PR DESCRIPTION
**Pull request by toteki**
- A bunch of tests
- One major bugfix to merge before release
- Additional edge case fixes

### CODE REVIEW

File: x/leverage/types/documented_test.go [hunk 1]
```diff
@@ -42,14 +42,14 @@ func TestMaxBorrowScenarioA(t *testing.T) {
 	assert.NilError(t, err)
 	assert.Equal(t,
 		"special:\n"+
-			"  40.000000000000000000AAAA, 20.000000000000000000BBBB, 0.500000000000000000\n"+
-			"  0.000000000000000000BBBB, 0.000000000000000000AAAA, 0.500000000000000000\n"+
-			"  50.000000000000000000AAAA, 20.000000000000000000CCCC, 0.400000000000000000\n"+
-			"  0.000000000000000000CCCC, 0.000000000000000000AAAA, 0.400000000000000000\n"+
+			"  [40 AAAA, 20 BBBB] @ 0.5\n"+
```
Comment (toteki) [inline x/leverage/types/documented_test.go:45]: See added string methods

File: x/leverage/types/string.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+// String represents a WeightedDecCoin in the form coin @ weight
+// e.g. 10 uumee @ 0.35
+func (wdc WeightedDecCoin) String() string {
+	return sDecCoin(wdc.Asset) + " @ " + sDec(wdc.Weight)
+}
+
+// sDec formats a sdk.Dec as a string with no trailing zeroes after the decimal point,
+// omitting the decimal point as well for whole numbers
+func sDec(d sdk.Dec) string {
```
Comment (toteki) [inline x/leverage/types/string.go:48]: optional, and might want to move to `util/string` package if they ever need to be exported.

File: x/leverage/types/weighted_coin.go [hunk 3]
```diff
@@ -95,27 +93,46 @@ func (wsp WeightedSpecialPair) before(b WeightedSpecialPair) bool {
 }
 
 // Add returns the sum of a weightedDecCoins and an additional weightedDecCoin.
-// The result is sorted.
+// The result is sorted, has any duplicate denoms combined, and any zero amounts omitted.
```
Comment (toteki) [inline x/leverage/types/weighted_coin.go:96]: This wasn't causing any bugs since positions are created from borrower collateral and borrows, which are unique in state.

The improvement does make the struct more resilient to unintended behavior if it ever occurs though.

File: x/leverage/types/weighted_coin.go [hunk 4]
```diff
@@ -72,8 +70,8 @@ func (wnp WeightedNormalPair) before(b WeightedNormalPair) bool {
 		return wnp.Borrow.Weight.GT(b.Borrow.Weight)
 	}
 	if wnp.Collateral.Asset.Denom != b.Collateral.Asset.Denom {
-		// if both weight are equal, sort by collateral denom, alphabetical
-		return wnp.Collateral.Asset.Denom != b.Collateral.Asset.Denom
+		// if both weights are equal, sort by collateral denom, alphabetical
+		return wnp.Collateral.Asset.Denom < b.Collateral.Asset.Denom
```
Comment (toteki) [inline x/leverage/types/weighted_coin.go:74]: This could have created inconsistent ordering of assets - although only ones with identical weights, so it wouldn't have affected positions. Still a bug though.

File: x/leverage/types/weighted_coin.go [hunk 5]
```diff
@@ -135,6 +152,7 @@ func (wdc WeightedDecCoins) Total(denom string) sdk.Dec {
 }
 
 // Sub subtracts a sdk.DecCoin from a WeightedDecCoins. Panics if the result would be negative.
+// Does not sort or change the order of the input denoms, combine duplicates, or remove zero coins.
```
Comment (toteki) [inline x/leverage/types/weighted_coin.go:155]: I think there are cases where this would have been a significant bug. Its fixes should be merged before release.

File: x/leverage/types/weighted_coin.go [hunk 6]
```diff
@@ -95,27 +93,46 @@ func (wsp WeightedSpecialPair) before(b WeightedSpecialPair) bool {
...
 		}
 	}
-	if !found {
-		sum = append(sum, add)
+	// collect total coins into a slice
+	for _, denom := range denoms {
+		sum = append(sum, weightedCoins[denom])
 	}
-	// sorts the sum. Fixes unsorted input as well.
+	// sorts the sum.
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/types/weighted_coin.go:135]: those optimizations assume well-formed input (pre-sorted, no duplicates) but this method is meant to fix bad input when encountered

File: x/leverage/types/string.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (ap *AccountPosition) String() string {
+	s := "special:\n"
+	for _, wsp := range ap.specialPairs {
+		s += fmt.Sprintf("  %s\n", wsp)
+	}
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/types/string.go:14]: One line per component would group all special pairs and all normal pairs into one line each - would be very crowded and hard to read.

File: x/leverage/types/weighted_coin.go [hunk 8]
```diff
@@ -148,19 +166,15 @@ func (wdc WeightedDecCoins) Sub(sub sdk.DecCoin) (diff WeightedDecCoins) {
 			diff = append(diff, c)
 		}
 	}
-	if !found {
+	if sub.IsPositive() && !found {
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/types/weighted_coin.go]: fixed

File: x/leverage/types/string.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+// String represents a WeightedNormalPair in the form [WeightedDecCoin, WeightedDecCoin]
+// e.g. [10 uumee @ 0.35, 3.5 uumee @ 0.35]
+func (wnp WeightedNormalPair) String() string {
+	return fmt.Sprintf("[%s, %s]", wnp.Collateral, wnp.Borrow)
+}
+
+// String represents a WeightedSpecialPair in the form [DecCoin, DecCoin] @ weight
+// e.g. [10 uumee, 3.5 uumee] @ 0.35
+func (wsp WeightedSpecialPair) String() string {
+	return "[" + sDecCoin(wsp.Collateral) + ", " + sDecCoin(wsp.Borrow) + "] @ " + sDec(wsp.SpecialWeight)
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/types/string.go]: fixed

File: x/leverage/types/string.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// String represents a WeightedNormalPair in the form [WeightedDecCoin, WeightedDecCoin]
+// e.g. [10 uumee @ 0.35, 3.5 uumee @ 0.35]
+func (wnp WeightedNormalPair) String() string {
+	return fmt.Sprintf("[%s, %s]", wnp.Collateral, wnp.Borrow)
+}
+
+// String represents a WeightedSpecialPair in the form [DecCoin, DecCoin] @ weight
+// e.g. [10 uumee, 3.5 uumee] @ 0.35
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/types/string.go]: fixed (though I put weight first so it's more clear that they're sorted by weight when listed together)

File: x/leverage/types/string.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	// otherwise, return with decimal intact but trailing zeroes removed
+	return parts[0] + "." + parts[1]
+}
+
+// sDecCoin formats a sdk.DecCoin with no trailing zeroes after the decimal point in its amount,
+// omitting the decimal point as well for whole numbers
+func sDecCoin(c sdk.DecCoin) string {
+	return sDec(c.Amount) + " " + c.Denom
+}
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/types/string.go:52]: done (but in `sdkutil` since the functions moved)

The asset pair & weighted coin stringers are tested indirectly by documented_test.go

**Comment by toteki** [comment]
Fixed suggestions. We need to merge this before release since it contains a fix @robert-zaremba

File: x/leverage/types/weighted_coin_test.go [hunk 12]
```diff
@@ -0,0 +1,1061 @@
...
+			Collateral: weightedDecCoin("WWWW", "1.0", "1.0"),
+			Borrow:     weightedDecCoin("WWWW", "1.0", "1.0"),
+		},
+		// this V -> A pair confirms that weight of the borrow (0.1)
+		// take precedence over alphabetical of the collateral (V-W)
+		// when weight of the collateral is equal (1.0)
+		{
+			Collateral: weightedDecCoin("VVVV", "1.0", "1.0"),
+			Borrow:     weightedDecCoin("AAAA", "1.0", "0.1"),
+		},
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/types/weighted_coin_test.go:425]: My IDE reverts this change whenever I try it
